# 对数几率回归

## 实验内容
1. 使用对数几率回归完成垃圾邮件分类问题和Dota2结果预测问题。
2. 计算十折交叉验证下的精度(accuracy)，查准率(precision)，查全率(recall)，F1值。

## 评测指标  
1. 精度
2. 查准率
3. 查全率
4. F1

# 1. 读取数据

In [1]:
import numpy as np

In [2]:
spambase = np.loadtxt('data/spambase/spambase.data', delimiter = ",")
dota2results = np.loadtxt('data/dota2Dataset/dota2Train.csv', delimiter=',')

# 2. 导入模型

In [3]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_predict

# 3. 提取数据

这里的spamx和dota2x包含了数据集内所有的特征

In [4]:
spamx = spambase[:, :57]
spamy = spambase[:, 57]

dota2x = dota2results[:, 1:]
dota2y = dota2results[:, 0]

# 4. 训练并预测

请你完成两个模型使用全部特征的训练与预测，并将预测结果存储起来

In [5]:
model = LogisticRegression()

y_pred = cross_val_predict(model, spamx, spamy, cv=10)



y_pred2 = cross_val_predict(model, dota2x, dota2y, cv=10)

# 5. 评价指标的计算

请你计算两个模型的四项指标

In [6]:
acc = accuracy_score(y_true=spamy, y_pred=y_pred)
recall = recall_score(y_true=spamy, y_pred=y_pred)
precision = precision_score(y_true=spamy, y_pred=y_pred)
f1 = f1_score(y_true=spamy, y_pred=y_pred)
print(acc, recall, precision, f1)


acc2 = accuracy_score(y_true=dota2y, y_pred=y_pred2)
recall2 = recall_score(y_true=dota2y, y_pred=y_pred2)
precision2 = precision_score(y_true=dota2y, y_pred=y_pred2)
f12 = f1_score(y_true=dota2y, y_pred=y_pred2)
print(acc2, recall2, precision2, f12)

0.9184959791349706 0.8869277440706013 0.9043869516310461 0.8955722639933167
0.598758769562871 0.6766225247017342 0.6066682595989487 0.6397387318415723


###### 双击此处填写

数据集|精度|查准率|查全率|F1
-|-|-|-|-
spambase | 0.9184959791349706  | 0.8869277440706013  | 0.9043869516310461  |0.8955722639933167
dota2Results | 0.598758769562871  | 0.6766225247017342 |0.6066682595989487  | 0.6397387318415723

# 6. 选做：尝试对特征进行变换、筛选、组合后，训练模型并计算十折交叉验证后的四项指标

In [7]:
from sklearn.preprocessing import StandardScaler,MinMaxScaler,Normalizer
def min_feature(x):
    return MinMaxScaler().fit_transform(x)


def nor_feature(x):
    return Normalizer().fit_transform(x)

def standard_feature(x):
    return StandardScaler().fit_transform(x)

spamx=min_feature(spamx)
y_pred = cross_val_predict(model, spamx, spamy, cv=10)
acc = accuracy_score(y_true=spamy, y_pred=y_pred)
recall = recall_score(y_true=spamy, y_pred=y_pred)
precision = precision_score(y_true=spamy, y_pred=y_pred)
f1 = f1_score(y_true=spamy, y_pred=y_pred)
print(acc, recall, precision, f1)

spamx=nor_feature(spamx)
y_pred = cross_val_predict(model, spamx, spamy, cv=10)
acc = accuracy_score(y_true=spamy, y_pred=y_pred)
recall = recall_score(y_true=spamy, y_pred=y_pred)
precision = precision_score(y_true=spamy, y_pred=y_pred)
f1 = f1_score(y_true=spamy, y_pred=y_pred)
print(acc, recall, precision, f1)

spamx=standard_feature(spamx)
spamx=nor_feature(spamx)

y_pred = cross_val_predict(model, spamx, spamy, cv=10)
acc = accuracy_score(y_true=spamy, y_pred=y_pred)
recall = recall_score(y_true=spamy, y_pred=y_pred)
precision = precision_score(y_true=spamy, y_pred=y_pred)
f1 = f1_score(y_true=spamy, y_pred=y_pred)
print(acc, recall, precision, f1)

0.886329058900239 0.8075013789299503 0.8937728937728938 0.8484497247174733
0.9198000434688112 0.8935466078323221 0.9020044543429844 0.8977556109725686
0.924798956748533 0.8957528957528957 0.9118472768107805 0.9037284362826934


1. 模型1的处理流程：特征归一化

2. 模型2的处理流程：特征去均值

3. 模型3的处理流程:特征去均值同时特征正则化

特征经过正则化和去均值后效果最好

模型|数据集|精度|查准率|查全率|F1
-|-|-|-|-|-
模型1 | 数据集 | 0.886329058900239 0.8075013789299503 0.8937728937728938 0.8484497247174733

模型2 | 数据集 | 0.9198000434688112 0.8935466078323221 0.9020044543429844 0.8977556109725686

模型3 | 数据集 | 0.924798956748533 0.8957528957528957 0.9118472768107805 0.9037284362826934
